identity-service/src/notifications/trendingTrackProcessing.js

```diff
@@ -0,0 +1,172 @@
+
+const models = require('../models')
+const { logger } = require('../logging')
+const config = require('../config.js')
+const {
+  deviceType,
+  notificationTypes
+} = require('./constants')
+const { publish } = require('./notificationQueue')
+const {
+  decodeHashId,
+  shouldNotifyUser
+} = require('./utils')
+const { fetchNotificationMetadata } = require('./fetchNotificationMetadata')
+const {
+  notificationResponseMap,
+  notificationResponseTitleMap,
+  pushNotificationMessagesMap
+} = require('./formatNotificationMetadata')
+
+const notifDiscProv = config.get('notificationDiscoveryProvider')
+
+const TRENDING_TIME = Object.freeze({
+  DAY: 'day',
+  WEEK: 'week',
+  MONTH: 'month',
+  YEAR: 'YEAR'
+})
+
+const TRENDING_GENRE = Object.freeze({
+  ALL: 'all'
+})
+
+const getTimeGenreActionType = (time, genre) => `${time}:${genre}`
+
+// The minimum time in hrs between notifications
+const TRENDING_INTERVAL = 3
+
+// The highest rank for which a notification will be sent
+const MAX_TOP_TRACK_RANK = 10
+
+async function getTrendingTracks () {
+  try {
+    // The owner info is then used to target listenCount milestone notifications
+    let params = new URLSearchParams()
+    params.append('time', TRENDING_TIME.WEEK)
+    params.append('limit', MAX_TOP_TRACK_RANK)
+
+    const trendingTracksResponse = await axios({
+      method: 'get',
+      url: `${notifDiscProv}/v1/full/tracks/trending?time=week`,
+      params,
+      timeout: 10000
+    })
+    const trendingTracks = trendingTracksResponse.data.data.map((track, idx) => ({
+      trackId: decodeHashId(track.id),
+      rank: idx + 1,
+      userId: decodeHashId(track.user.id)
+    }))
+    const blocknumber = trendingTracksResponse.data.latest_indexed_block
+    return { trendingTracks, blocknumber }
+  } catch (err) {
+    console.log(err)
+    return null
+  }
+}
+
+async function processTrendingTracks (audiusLibs, blocknumber, trendingTracks, tx) {
+  const now = moment()
+  for (let idx = 0; idx < trendingTracks.length; idx += 1) {
+    const { rank, trackId, userId } = trendingTracks[idx]
+    const { notifyMobile, notifyBrowserPush } = await shouldNotifyUser(userId, 'milestonesAndAchievements')
+
+    // Check if the notification was previously created
+    let existingTrendingTracks = await models.Notification.findAll({
+      where: {
+        userId: userId,
+        type: notificationTypes.TrendingTrack,
+        entityId: trackId
+      },
+      include: [{
+        model: models.NotificationAction,
+        as: 'actions'
+      }],
+      order: [['timestamp', 'DESC']],
+      limit: 1,
+      transaction: tx
+    })
+    if (existingTrendingTracks.length > 0) {
+      const previousRank = existingTrendingTracks[0].actions[0].actionEntityId
+      const previousCreated = moment(existingTrendingTracks[0].timestamp)
+      const duration = moment.duration(now.diff(previousCreated)).asHours()
+      if (duration < TRENDING_INTERVAL || previousRank <= rank) {
```

Comment:
if previousRank is <= rank (incoming), shouldn't that actually trigger a notif?

Reply:
Rank is the position in the list. 
If the track was previously trending at 4 and either stayed the same or moved down to like 5 then the user does not get a notification. 
Only if the track was not previously trending or moved up the trending list from a lower to higher position 
ie. from rank 10 trending to rank 1 trending does the user get a notification. 
So if previousRank was 5 and current rank is 6 then  `previousRank <= rank` mean the user does not get a notification.

Reply:
@hareeshnagaraj this looks right to me - rank 1 (1st place) is higher than rank 2 (2nd place)
imo this could use a comment explaining it though, not super intuitive

Reply:
ah i see, this makes sense I had it switched mentally. let's say a track re-enters trending position 9 2 months after it was at trending position 3 - should that still trigger a notif? not saying it has to, jw if that's the intended behavior

Reply:
nvm your comment below explains this